File: StateMaschine.kt
package de.a4evar.statemaschine

import kotlin.coroutines.*
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel


typealias Condition = (StateMaschine.Transition) -> Boolean

typealias Action = (StateMaschine.Transition) -> Unit

interface State
interface Event


class StateMaschine(val context: CoroutineContext = Dispatchers.Default) :
  CoroutineScope by CoroutineScope(Dispatchers.Default) {

  val DEBUG = false

  companion object {
    val STATE_MASCHINE_START = object : Event {
      override fun toString(): String {
        return "STATE_MASCHINE_START"
      }
    }
    val STATE_MASCHINE_INITIAL = object : State {
      override fun toString(): String {
        return "STATE_MASCHINE_INITIAL"
      }
    }
  }

  private val stateNodes = mutableListOf<StateNode>()

  private var rootStateNode: StateNode? = null

  private var currentStateNode: StateNode? = null

  private val eventActions = mutableListOf<EventAction>()

  private val channel = Channel<Event>(Channel.UNLIMITED)

  val currentState
    get() = currentStateNode?.state


  fun addTransition(
    fromState: State,
    toState: State,
    vararg events: Event,
    condition: Condition? = null
  ) {
    if (events.isEmpty()) throw Exception("must specify an event")

    for (event in events) {
      var fromStateNode = stateNodes.find { it.state == fromState }
      if (fromStateNode == null) {
        if (rootStateNode == null) { // first transition added
          fromStateNode = StateNode(fromState)
          stateNodes.add(fromStateNode)
          rootStateNode = fromStateNode
        } else {
          throw Exception("fromState not found")
        }
      }

//      if (fromStateNode.events[event] != null) {
//        throw Exception("transition for event $event from state $fromState exists already")
//      }

      val transitions = fromStateNode.events[event]
      if (transitions == null) {
        fromStateNode.events[event] =
          mutableListOf(Transition(fromState, toState, event, condition))
      } else {
        transitions.add(Transition(fromState, toState, event, condition))
      }

      val toStateNode = stateNodes.find { it.state == toState }
      if (toStateNode == null) {
        stateNodes.add(StateNode(toState))
      }
    }
  }


  fun onDepartureFrom(
    fromState: State,
    toState: State? = null,
    event: Event? = null,
    action: Action
  ) {
    val fromStateNode =
      stateNodes.find { it.state == fromState } ?: throw Exception("state not found: $fromState")
    fromStateNode.departureActions.add(DepartureAction(fromState, toState, event, action))
  }


  fun onArrivalAt(toState: State, fromState: State? = null, event: Event? = null, action: Action) {
    val toStateNode =
      stateNodes.find { it.state == toState } ?: throw Exception("state not found: $toState")
    toStateNode.arrivalActions.add(ArrivalAction(toState, fromState, event, action))
  }


  fun onEvent(event: Event, fromState: State? = null, toState: State? = null, action: Action) {
    eventActions.add(EventAction(event, fromState, toState, action))
  }


  fun start(startState: State? = null) {
    if (currentStateNode != null) throw Exception("state maschine already running")
    if (startState != null) {
      currentStateNode = stateNodes.find { it.state == startState }
        ?: throw Exception("state not found: $startState")
    } else {
      currentStateNode = rootStateNode ?: throw Exception("state maschine not configured")
    }

    launch {
      try {
        for (action in currentStateNode!!.arrivalActions) {
          if (action.event == STATE_MASCHINE_START) {
            if (DEBUG) println("executing arrival action for event STATE_MASCHINE_START")
            action.execute(Transition(STATE_MASCHINE_INITIAL, currentState!!, STATE_MASCHINE_START))?.join()
            break
          }
        }

        for (event in channel) {
          if (DEBUG) println("received event $event")
          val csn = currentStateNode ?: throw Exception("state maschine not started")
          val transitions = csn.events[event] ?: throw Exception("cannot find transition")
          for (transition in transitions) {
            if (transition.isValid()) {
              if (DEBUG) println("transitioning to state ${transition.toState}")
              for (action in eventActions.filter { it.event == event }) {
                action.execute(transition)?.join()
              }

              for (action in csn.departureActions) {
                action.execute(transition)?.join()
              }

              currentStateNode = stateNodes.find { it.state == transition.toState }

              for (action in currentStateNode!!.arrivalActions) {
                action.execute(transition)?.join()
              }

              break
            }
          }
        }
      } catch (e: Throwable) {
        println(e.localizedMessage)
        throw e
      }
    }

    if (DEBUG) println("state maschine started")
  }


  fun stop() {
    channel.close()
    cancel()
  }


  infix fun signal(event: Event) {
    currentStateNode ?: throw Exception("state maschine not started")
    channel.offer(event)
  }


  inner class StateNode(val state: State) {
    internal val events = mutableMapOf<Event, MutableList<Transition>>()
    internal val departureActions = mutableListOf<DepartureAction>()
    internal val arrivalActions = mutableListOf<ArrivalAction>()
    override fun toString() = state.toString()
  }


  inner class Transition(
    val fromState: State,
    val toState: State,
    val event: Event,
    val condition: Condition? = null
  ) {
    fun isValid() = condition?.invoke(this) ?: true
    override fun toString(): String {
      return "Transition $fromState ---$event--> $toState"
    }
  }


  abstract inner class TransitionAction(val action: Action) {
    abstract suspend fun execute(transition: Transition): Job?
  }

  inner class DepartureAction(
    val fromState: State,
    val toState: State? = null,
    val event: Event? = null,
    action: Action
  ) : TransitionAction(action) {
    override suspend fun execute(transition: Transition): Job? {
      if ((toState == null || toState == transition.toState) && (event == null || event == transition.event)) {
        return launch(context) {
          action(transition)
        }
      } else {
        return null
      }
    }
  }

  inner class ArrivalAction(
    val toState: State,
    val fromState: State? = null,
    val event: Event? = null,
    action: Action
  ) :
    TransitionAction(action) {
    override suspend fun execute(transition: Transition): Job? {
      if ((fromState == null || fromState == transition.fromState) && (event == null || event == transition.event)) {
        return launch(context) {
          action(transition)
        }
      } else {
        return null
      }
    }
  }

  inner class EventAction(
    val event: Event,
    val fromState: State? = null,
    val toState: State? = null,
    action: Action
  ) :
    TransitionAction(action) {
    override suspend fun execute(transition: Transition): Job? {
      if ((toState == null || toState == transition.toState) && (fromState == null || fromState == transition.fromState)) {
        return launch(context) {
          action(transition)
        }
      } else {
        return null
      }
    }
  }


  internal var _tt: TransitionTemplate? = null

  inner class TransitionTemplate(
    var fromState: State? = null,
    var toState: State? = null,
    var event: Event? = null,
    var condition: Condition? = null
  ) {
    fun generate() {
      addTransition(
        fromState ?: throw Exception("incomplete transition (from)"),
        toState ?: throw Exception("incomplete transition (to)"),
        event ?: throw Exception("incomplete transition (via)"),
        condition = condition
      )
      _tt = null
    }

  }

  val transition: TransitionTemplate
    get() {
      _tt?.generate()
      _tt = TransitionTemplate()
      return _tt!!
    }

  infix fun TransitionTemplate.from(state: State): TransitionTemplate {
    this.fromState = state; return this
  }

  infix fun TransitionTemplate.to(state: State): TransitionTemplate {
    this.toState = state; return this
  }

  infix fun TransitionTemplate.via(event: Event): TransitionTemplate {
    this.event = event; return this
  }

  infix fun TransitionTemplate.by(event: Event) = via(event)

  infix fun TransitionTemplate.check(condition: Condition): TransitionTemplate {
    this.condition = condition; return this
  }


  enum class ActionKind { ON_ARRIVAL, ON_DEPARTURE, ON_EVENT }

  internal var _at: ActionTemplate? = null

  inner class ActionTemplate(
    private val kind: ActionKind,
    var fromState: State? = null,
    var toState: State? = null,
    var event: Event? = null
  ) {

    fun generate(action: Action) {
      _tt?.generate()

      when (kind) {
        ActionKind.ON_ARRIVAL -> onArrivalAt(
          toState ?: throw Exception("incomplete action (at)"),
          fromState,
          event,
          action
        )
        ActionKind.ON_DEPARTURE -> onDepartureFrom(
          fromState ?: throw Exception("incomplete action (from)"),
          toState,
          event,
          action
        )
        ActionKind.ON_EVENT -> onEvent(
          event ?: throw Exception("incomplete action (from)"),
          toState,
          fromState,
          action
        )
      }

      _at = null
    }
  }

  val arriving: ActionTemplate
    get() {
      if (_at != null) throw Exception("previous action is missing run block")
      _at = ActionTemplate(ActionKind.ON_ARRIVAL)
      return _at!!
    }
  val departing: ActionTemplate
    get() {
      if (_at != null) throw Exception("previous action is missing run block")
      _at = ActionTemplate(ActionKind.ON_DEPARTURE)
      return _at!!
    }
  val receiving: ActionTemplate
    get() {
      if (_at != null) throw Exception("previous action is missing run block")
      _at = ActionTemplate(ActionKind.ON_EVENT)
      return _at!!
    }

  infix fun ActionTemplate.at(state: State): ActionTemplate {
    this.toState = state; return this
  }

  infix fun ActionTemplate.to(state: State) = at(state)

  infix fun ActionTemplate.from(state: State): ActionTemplate {
    this.fromState = state; return this
  }

  infix fun ActionTemplate.via(event: Event): ActionTemplate {
    this.event = event; return this
  }

  infix fun ActionTemplate.of(event: Event) = via(event)

  infix fun ActionTemplate.by(event: Event) = via(event)

  infix fun ActionTemplate.run(action: Action) {
    generate(action)
  }

}


fun StateMaschine(context: CoroutineContext = Dispatchers.Default, init: StateMaschine.() -> Unit): StateMaschine {
  val sm = StateMaschine(context)
  sm.init()
  sm._tt?.generate()
  if (sm._at != null) throw Exception("last action is missing run block")
  return sm
}

File: StateMaschine.swift

protocol State: Equatable {
  var name: String { get }
}

extension State {
  var debugDescription: String { get { return "state \(name)" } }
}

protocol Event: Hashable {
  var name: String { get }
}

extension Event {
  var debugDescription: String { get { return "event \(name)" } }
}

protocol TransitionAction {
  associatedtype S: State
  associatedtype E: Event
  func execute(transition: StateMaschine<S, E>.Transition)
}

fileprivate let DEBUG = true

enum StateMaschineErrors: Error {
  case GeneralError(_ message: String)
}


class StateMaschine<S: State, E: Event> {
  
  typealias Condition = (Transition) -> Bool

  typealias Action = (Transition) -> Void

  private var stateNodes = [StateNode]()
  
  private var rootStateNode: StateNode?
  
  private var currentStateNode: StateNode?
  
  private var eventActions = [EventAction]()
  
  var currentState: S? {
    get {
      return currentStateNode?.state
    }
  }
  
  var eventQueue: DispatchQueue?
  
  func addTransition(fromState: S, toState: S, events: E..., condition: Condition? = nil) throws {
    if events.isEmpty { throw StateMaschineErrors.GeneralError("must specify an event") }
    
    for event in events {
      var fromStateNode = stateNodes.first { return $0.state == fromState }
      if fromStateNode == nil {
        if rootStateNode != nil { throw StateMaschineErrors.GeneralError("fromState not found") }
        
        fromStateNode = StateNode(fromState)
        stateNodes.append(fromStateNode!)
        rootStateNode = fromStateNode
      }
      
      let transition = Transition(fromState, toState, event, condition)
      var transitions = fromStateNode?.events[event]
      if transitions == nil {
        fromStateNode?.events[event] = [transition]
      } else {
        transitions?.append(transition)
      }
      
      let toStateNode = stateNodes.first { $0.state == toState }
      if toStateNode == nil {
        stateNodes.append(StateNode(toState))
      }
    }
  }
  
  
  func onDepartureFrom(fromState: S, toState: S?, event: E?, action: @escaping Action) throws {
    guard let fromStateNode = stateNodes.first(where: { $0.state == fromState }) else {
      throw StateMaschineErrors.GeneralError("state not found: \(fromState.name)")
    }
    fromStateNode.departureActions.append(DepartureAction(fromState, toState, event, action))
  }
  
  func onArrivalAt(toState: S, fromState: S?, event: E?, action: @escaping Action) throws {
    guard let toStateNode = stateNodes.first(where: { $0.state == toState }) else {
      throw StateMaschineErrors.GeneralError("state not found: \(toState.name)")
    }
    toStateNode.arrivalActions.append(ArrivalAction(toState, fromState, event, action))
  }
  
  func onEvent(event: E, fromState: S?, toState: S?, action: @escaping Action) throws {
    eventActions.append(EventAction(event, fromState, toState, action))
  }
  
  
  func start(startState: S? = nil) throws {
    guard (currentStateNode == nil) else {
      throw StateMaschineErrors.GeneralError("state machine already running")
    }
    
    if let startState = startState {
      currentStateNode = stateNodes.first { $0.state == startState }
      guard currentStateNode != nil else {
        throw StateMaschineErrors.GeneralError("start state not found: \(startState)")
      }
    } else {
      currentStateNode = rootStateNode
      guard currentStateNode != nil else {
        throw StateMaschineErrors.GeneralError("state machine not configured")
      }
    }
    
    self.eventQueue = DispatchQueue(label: "StateMaschineEvents", qos: .background)

    if DEBUG { print("state maschine started") }
  }
  
  
  func signal(event: E) {
    if DEBUG { print("received event \(event)") }
    eventQueue?.async {
      if let csn = self.currentStateNode {
        if let transitions = csn.events[event] {
          for transition in transitions {
            if transition.isValid() {
              if DEBUG { print("transitioning to state \(transition.toState)") }
              for action in self.eventActions.filter({ $0.event == event }) {
                action.execute(transition: transition)
              }
              
              for action in csn.departureActions {
                action.execute(transition: transition)
              }
              
              self.currentStateNode = self.stateNodes.first { $0.state == transition.toState }
              
              for action in self.currentStateNode!.arrivalActions {
                action.execute(transition: transition)
              }
              
              break // only process the first valid transition
            }
          }
        }
      }
    }
  }
  
  static func >>> (lhs: E, rhs: StateMaschine<S, E>) {
    rhs.signal(event: lhs)
  }
  
  class StateNode: Equatable {
    let state: S
    init(_ state: S) {
      self.state = state
    }
    var events = [E:[Transition]]()
    var departureActions = [DepartureAction]()
    var arrivalActions = [ArrivalAction]()
    
    static func == (lhs: StateNode, rhs: StateNode) -> Bool {
      return lhs.state == rhs.state
    }
  }
  
  
  class Transition {
    let fromState: S
    let toState: S
    let event: E
    let condition: Condition?
    
    init(_ fromState: S, _ toState: S, _ event: E, _ condition: Condition? = nil) {
      self.toState = toState
      self.fromState = fromState
      self.event = event
      self.condition = condition
    }
    
    func isValid() -> Bool {
      return condition?(self) ?? true
    }
  }
  
  
  class DepartureAction: TransitionAction {
    let fromState: S
    let toState: S?
    let event: E?
    var action: Action
    
    init(_ fromState: S, _ toState: S? = nil, _ event: E? = nil, _ action: @escaping Action) {
      self.fromState = fromState
      self.toState = toState
      self.event = event
      self.action = action
    }
    
    func execute(transition: Transition) {
      if ((toState == nil || toState?.name == transition.toState.name) &&
        (event == nil || event == transition.event)) {
        action(transition)
      }
    }
  }
  
  
  class ArrivalAction: TransitionAction {
    let fromState: S?
    let toState: S
    let event: E?
    var action: Action

    init(_ toState: S, _ fromState: S? = nil, _ event: E? = nil, _ action: @escaping Action) {
      self.fromState = fromState
      self.toState = toState
      self.event = event
      self.action = action
    }
    
    func execute(transition: Transition) {
      if ((fromState == nil || fromState == transition.fromState) &&
        (event == nil || event == transition.event)) {
        action(transition)
      }
    }
  }
  
  
  class EventAction: TransitionAction {
    let fromState: S?
    let toState: S?
    let event: E
    var action: Action

    init(_ event: E, _ fromState: S? = nil, _ toState: S? = nil, _ action: @escaping Action) {
      self.fromState = fromState
      self.toState = toState
      self.event = event
      self.action = action
    }
    
    func execute(transition: Transition) {
      if ((toState == nil || toState == transition.toState) &&
        (fromState == nil || fromState == transition.fromState)) {
        action(transition)
      }
    }
  }


  class TransitionTemplate {
    let sm: StateMaschine<S, E>
    var fromState: S?
    var toState: S?
    var event: E?
    var condition: Condition?
    
    init(sm: StateMaschine<S, E>, fromState: S) {
      self.sm = sm
      self.fromState = fromState
    }
    
    @discardableResult
    static func >>> (lhs: TransitionTemplate, rhs: E) -> TransitionTemplate {
      lhs.event = rhs
      return lhs
    }
    
    @discardableResult
    static func >>> (lhs: TransitionTemplate, rhs: S) -> TransitionTemplate {
      lhs.toState = rhs
      return lhs
    }
    
    static func ??? (lhs: TransitionTemplate, rhs: Condition?) {
      lhs.condition = rhs
      lhs.generate()
    }
    
    func generate() {
      try? sm.addTransition(fromState: fromState!, toState: toState!, events: event!, condition: condition)
    }
  }
  
  static func =+ (lhs: StateMaschine<S, E>, rhs: S) -> TransitionTemplate {
    return TransitionTemplate(sm: lhs, fromState: rhs)
  }
  

  class ActionTemplate {
    enum ActionKind { case ON_ARRIVAL, ON_DEPARTURE, ON_EVENT }
    
    let kind: ActionKind
    let sm: StateMaschine<S, E>
    var fromState: S?
    var toState: S?
    var event: E?
    var action: Action?

    init(sm: StateMaschine<S, E>, kind: ActionKind, fromState: S? = nil, toState: S? = nil, event: E? = nil) {
      self.sm = sm
      self.kind = kind
      self.fromState = fromState
      self.toState = toState
      self.event = event
    }
    
    @discardableResult
    static func >>> (lhs: ActionTemplate, rhs: S) -> ActionTemplate {
      if lhs.toState == nil {
        lhs.toState = rhs
      } else {
        lhs.fromState = rhs
      }
      return lhs
    }
    
    @discardableResult
    static func <<< (lhs: ActionTemplate, rhs: S) -> ActionTemplate {
      lhs.fromState = rhs
      return lhs
    }
    
    @discardableResult
    static func >>> (lhs: ActionTemplate, rhs: E) -> ActionTemplate {
      lhs.event = rhs
      return lhs
    }
    
    static func !!! (lhs: ActionTemplate, rhs: @escaping Action) {
      lhs.action = rhs
      lhs.generate(action: rhs)
    }
    
    func generate(action: @escaping Action) {
      switch kind {
      case .ON_DEPARTURE:
        try? sm.onDepartureFrom(fromState: fromState!, toState: toState, event: event, action: action)
      case .ON_ARRIVAL:
        try? sm.onArrivalAt(toState: toState!, fromState: fromState, event: event, action: action)
      case .ON_EVENT:
        try? sm.onEvent(event: event!, fromState: fromState, toState: toState, action: action)
      }
    }
  }
  
  static func => (lhs: StateMaschine<S, E>, rhs: S) -> ActionTemplate {
    return ActionTemplate(sm: lhs, kind: .ON_DEPARTURE, fromState: rhs)
  }
  
  static func =< (lhs: StateMaschine<S, E>, rhs: S) -> ActionTemplate {
    return ActionTemplate(sm: lhs, kind: .ON_ARRIVAL, toState: rhs)
  }
  
  static func =! (lhs: StateMaschine<S, E>, rhs: E) -> ActionTemplate {
    return ActionTemplate(sm: lhs, kind: .ON_EVENT, event: rhs)
  }
}

infix operator =+  : AdditionPrecedence
infix operator >>> : AdditionPrecedence
infix operator <<< : AdditionPrecedence
infix operator ??? : AdditionPrecedence
infix operator !!! : AdditionPrecedence
infix operator =>  : AdditionPrecedence
infix operator =<  : AdditionPrecedence
infix operator =!  : AdditionPrecedence

func foobar() {
  
  enum TestEvent: String, Event {
    var name: String {
      get { return self.rawValue }
    }
    
    case START, STOP
  }
  
  enum TestState: String, State {
    var name: String {
      get { return self.rawValue }
    }
    
    case STOPPED, STARTED
  }
  
  let sm = StateMaschine<TestState, TestEvent>()
  sm =+ TestState.STARTED >>> TestEvent.START >>> TestState.STOPPED ??? { _ in return true }
  sm =+ TestState.STARTED >>> TestEvent.STOP >>> TestState.STOPPED ??? nil
  
  sm => TestState.STOPPED >>> TestEvent.START >>> TestState.STARTED !!! { _ in }
  sm =< TestState.STARTED !!! { _ in }
  sm =! TestEvent.START !!! { _ in }
 
}


/*
 
 
 protocol State {
 var name: String { get }
 }
 
 protocol Event {
 var name: String { get }
 }
 
 typealias Condition = (Transition) -> Bool
 
 typealias Action = (Transition) -> Void
 
 enum StateMaschineErrors: Error {
 case GeneralError(String)
 }
 
 
 class StateMaschine {
 
 class STATE_MASCHINE_START_CLASS: Event {
 var name: String {
 get {
 return "STATE_MASCHINE_START"
 }
 }
 }
 
 static let STATE_MASCHINE_START: Event = STATE_MASCHINE_START_CLASS()
 
 class STATE_MASCHINE_INITIAL_CLASS: State {
 var name: String {
 get {
 return "STATE_MASCHINE_INITIAL"
 }
 }
 }
 
 static let STATE_MASCHINE_INITIAL: State = STATE_MASCHINE_INITIAL_CLASS()
 
 private var stateNodes = [StateNode]()
 
 private var rootStateNode: StateNode?
 
 private var currentStateNode: StateNode?
 
 private var eventActions = [EventAction]()
 
 var currentState: State? {
 get {
 return currentStateNode?.state
 }
 }
 
 func addTransition(fromState: State, toState: State, events: Event..., condition: Condition? = nil) throws {
 if events.isEmpty { throw StateMaschineErrors.GeneralError("must specify an event") }
 
 for event in events {
 var fromStateNode = stateNodes.first { return $0.state.name == fromState.name }
 if fromStateNode == nil {
 if rootStateNode != nil { throw StateMaschineErrors.GeneralError("fromState not found") }
 
 fromStateNode = StateNode(fromState)
 stateNodes.append(fromStateNode!)
 rootStateNode = fromStateNode
 }
 
 let transition = Transition(fromState, toState, event, condition)
 var transitions = fromStateNode?.events[event.name]
 if transitions == nil {
 fromStateNode?.events[event.name] = [transition]
 } else {
 transitions?.append(transition)
 }
 
 let toStateNode = stateNodes.first { $0.state.name == toState.name }
 if toStateNode == nil {
 stateNodes.append(StateNode(toState))
 }
 }
 }
 
 
 func onDepartureFrom(fromState: State, toState: State?, event: Event?, action: @escaping Action) throws {
 guard var fromStateNode = stateNodes.first(where: { $0.state.name == fromState.name }) else {
 throw StateMaschineErrors.GeneralError("state not found: \(fromState.name)")
 }
 fromStateNode.departureActions.append(DepartureAction(fromState, toState, event, action))
 }
 
 func onArrivalAt(toState: State, fromState: State?, event: Event?, action: @escaping Action) throws {
 guard var toStateNode = stateNodes.first(where: { $0.state.name == toState.name }) else {
 throw StateMaschineErrors.GeneralError("state not found: \(toState.name)")
 }
 toStateNode.arrivalActions.append(ArrivalAction(toState, fromState, event, action))
 }
 
 func onEvent(event: Event, fromState: State?, toState: State?, action: @escaping Action) throws {
 eventActions.append(EventAction(event, fromState, toState, action))
 }
 
 
 
 
 }
 
 
 struct StateNode: Equatable {
 let state: State
 init(_ state: State) {
 self.state = state
 }
 var events = [String:[Transition]]()
 fileprivate var departureActions = [DepartureAction]()
 fileprivate var arrivalActions = [ArrivalAction]()
 
 static func == (lhs: StateNode, rhs: StateNode) -> Bool {
 return lhs.state.name == rhs.state.name
 }
 }
 
 
 struct Transition {
 let fromState: State
 let toState: State
 let event: Event
 let condition: Condition?
 
 init(_ fromState: State, _ toState: State, _ event: Event, _ condition: Condition? = nil) {
 self.toState = toState
 self.fromState = fromState
 self.event = event
 self.condition = condition
 }
 
 func isValid() -> Bool {
 return condition?(self) ?? true
 }
 }
 
 
 protocol TransitionAction {
 func execute(transition: Transition)
 }
 
 private class AbstractTransitionAction {
 let fromState: State?
 let toState: State?
 let event: Event?
 var action: Action
 
 init(_ fromState: State?, _ toState: State? = nil, _ event: Event? = nil, _ action: @escaping Action) {
 self.fromState = fromState
 self.toState = toState
 self.event = event
 self.action = action
 }
 }
 
 private class DepartureAction: AbstractTransitionAction, TransitionAction {
 
 init(_ fromState: State, _ toState: State? = nil, _ event: Event? = nil, _ action: @escaping Action) {
 super.init(fromState, toState, event, action)
 }
 
 func execute(transition: Transition) {
 if ((toState == nil || toState?.name == transition.toState.name) &&
 (event == nil || event?.name == transition.event.name)) {
 action(transition)
 }
 }
 }
 
 
 private class ArrivalAction: AbstractTransitionAction {
 
 init(_ toState: State, _ fromState: State? = nil, _ event: Event? = nil, _ action: @escaping Action) {
 super.init(fromState, toState, event, action)
 }
 
 func execute(transition: Transition) {
 if ((fromState == nil || fromState?.name == transition.fromState.name) &&
 (event == nil || event?.name == transition.event.name)) {
 action(transition)
 }
 }
 }
 
 
 private class EventAction: AbstractTransitionAction {
 
 init(_ event: Event, _ fromState: State? = nil, _ toState: State? = nil, _ action: @escaping Action) {
 super.init(fromState: fromState, toState: toState, event: event, action: action)
 }
 
 func execute(transition: Transition) {
 if ((toState == nil || toState?.name == transition.toState.name) &&
 (fromState == nil || fromState?.name == transition.fromState.name)) {
 action(transition)
 }
 }
 }
 
 
 */
